server/backends/blobstore/aws/aws.go

```diff
@@ -254,3 +257,53 @@ func (a *AwsS3BlobStore) BlobExists(ctx context.Context, blobName string) (bool,
 
 	return true, nil
 }
+
+func (a *AwsS3BlobStore) Writer(ctx context.Context, blobName string) (interfaces.CommittedWriteCloser, error) {
+	// Open a pipe.
+	pr, pw := io.Pipe()
+
+	errch := make(chan error, 1)
+
+	ctx, cancel := context.WithCancel(ctx)
+
+	// Upload from pr in a separate Go routine.
+	go func() {
+		defer pr.Close()
+		_, err := a.uploader.UploadWithContext(
+			ctx,
+			&s3manager.UploadInput{
+				Bucket: a.bucket,
+				Key:    aws.String(util.BlobPath(blobName)),
+				Body:   pr,
+			},
+		)
+		errch <- err
+		close(errch)
+	}()
+
+	zw := util.NewCompressWriter(pw)
+	cwc := ioutil.NewCustomCommitWriteCloser(zw)
+	cwc.CommitFn = func(int64) error {
+		if compresserCloseErr := zw.Close(); compresserCloseErr != nil {
+			cancel() // Don't try to finish the commit op if Close() failed.
+			if pipeCloseErr := pw.Close(); pipeCloseErr != nil {
+				log.Errorf("Error closing the pipe for %s: %s", blobName, pipeCloseErr)
+			}
+			<-errch // Canceling the context makes any error here meaningless.
+			return compresserCloseErr
+		}
+		var err error
+		if writerCloseErr := pw.Close(); writerCloseErr != nil {
+			err = writerCloseErr
```

Comment:
You can return the error directly instead of storing it. Since errch is buffered, there's no harm if you return before reading from the channel.